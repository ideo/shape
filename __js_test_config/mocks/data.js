import _ from 'lodash'
import { observable } from 'mobx'

const fakeJsonApiAttrs = {
  assign: jest.fn(),
  assignRef: jest.fn(),
  save: jest.fn().mockReturnValue(Promise.resolve({})),
  create: jest.fn().mockReturnValue(Promise.resolve({})),
  patch: jest.fn().mockReturnValue(Promise.resolve({})),
  update: jest.fn().mockReturnValue(Promise.resolve({})),
  disableMenu: jest.fn(),
}
export const fakeCollectionCard = {
  id: '11',
  order: 1,
  height: 1,
  width: 1,
  row: 0,
  col: 1,
  maxWidth: 1,
  maxHeight: 1,
  record: {},
  item: {},
  reference: false,
  image_contain: false,
  section_type: null,
  beginReplacing: jest.fn(),
  reselectOnlyEditableRecords: jest.fn(),
  reselectOnlyMovableCards: jest.fn(),
  API_create: jest.fn(),
  API_archive: jest.fn(),
  API_linkToMyCollection: jest.fn(),
  ...fakeJsonApiAttrs,
}

export const fakeQuillOp = { insert: 'hello world \n' }
export const fakeQuillData = { ops: [fakeQuillOp] }

export const fakeTextItemAttrs = {
  id: '1',
  type: 'Item::TextItem',
  internalType: 'items',
  name: 'My Cool Item',
  can_view: true,
  quill_data: fakeQuillData,
  breadcrumb: [['collections', 1, 'Some collection'], ['items', 1, 'my item']],
  parentPath: '/',
  can_edit: false,
  inherited_tag_list: [],
  internalType: 'items',
  fullyLoaded: true,
  parent_collection_card: { ...fakeCollectionCard },
}

const areaChartData = [
  { date: '2018-07-10', value: 10 },
  { date: '2018-08-10', value: 25 },
  { date: '2018-09-10', value: 30 },
]

export const fakeAreaChartDataset = {
  identifier: 'question',
  measure: 'participants',
  description: 'A description',
  timeframe: 'month',
  chart_type: 'area',
  order: 0,
  tiers: [],
  data: areaChartData,
  dataWithDates: areaChartData.map(d => ({
    date: new Date(d.date),
    value: d.value,
  })),
}

const fakeBarChartData = [
  { column: 1, value: 5, percentage: 50, type: 'question_context' },
  { column: 2, value: 0, percentage: 0, type: 'question_context' },
  { column: 3, value: 5, percentage: 50, type: 'question_context' },
  { column: 4, value: 0, percentage: 0, type: 'question_context' },
]

export const fakeBarChartDataset = {
  identifier: 'question',
  measure: 'participants',
  description: 'A description',
  question_type: 'question_context',
  timeframe: 'month',
  chart_type: 'bar',
  order: 0,
  total: 10,
  max_domain: 95,
  tiers: [],
  data: fakeBarChartData,
  dataWithDates: fakeBarChartData,
  isEmojiOrScaleQuestion: jest.fn(),
}

export const fakeDataItemCollectionsItemsAttrs = {
  ...fakeTextItemAttrs,
  type: 'Item::DataItem',
  quill_data: null,
  report_type: 'report_type_collections_and_items',
  isReportTypeCollectionsItems: true,
  isReportTypeNetworkAppMetric: false,
  isReportTypeRecord: false,
  data_settings: {
    d_measure: 'participants',
    d_timeframe: 'month',
  },
  measure: {
    name: 'Participants',
  },
  primaryDataset: fakeAreaChartDataset,
  datasets: [fakeAreaChartDataset],
}

const creativeDifferenceData = [
  { date: '2018-07-10', value: 10 },
  { date: '2018-08-10', value: 25 },
  { date: '2018-09-10', value: 30 },
]

export const creativeDifferenceQualityDataset = {
  name: 'purpose',
  identifier: 'purpose',
  measure: 'Purpose',
  description:
    'The degree to which there is alignment about a meaningful change that leadership and employees want to make in the world.',
  timeframe: 'month',
  chart_type: 'area',
  single_value: 0,
  order: 0,
  style: {
    fill: '#EFEFEF',
    dashWidth: 2,
  },
  max_domain: 60,
  tiers: [
    { value: 0, name: 'Novice' },
    { value: 20, name: 'Learning' },
    { value: 40, name: 'Expert' },
  ],
  data: creativeDifferenceData,
  dataWithDates: creativeDifferenceData,
}

export const fakeDatasetAttrs = {
  chart_type: 'bar',
  data: [],
  data_items_datasets_id: 100,
  measure: 'participants',
  identifier: 'dataset',
  name: 'dataset',
  order: 0,
  question_type: null,
  selected: true,
  test_collection_id: null,
  timeframe: 'ever',
  total: 1,
  tiers: [],
}

export const fakeDataset = {
  ...fakeDatasetAttrs,
  rawAttributes: jest.fn().mockReturnValue(fakeDatasetAttrs),
  getRecordType: jest.fn().mockReturnValue('datasets'),
  toJSON: jest.fn().mockReturnValue(fakeDatasetAttrs),
  ...fakeJsonApiAttrs,
}

export const fakeDataItemRecordAttrs = {
  ...fakeTextItemAttrs,
  type: 'Item::DataItem',
  quill_data: null,
  datasets: [fakeDataset],
  name: 'Data Item',
  report_type: 'report_type_record',
  isReportTypeCollectionsItems: false,
  isReportTypeNetworkAppMetric: false,
  isReportTypeRecord: true,
  primaryDataset: creativeDifferenceQualityDataset,
  datasets: [
    creativeDifferenceQualityDataset,
    {
      ...creativeDifferenceQualityDataset,
      measure: '95th Percentile',
      order: 1,
      chart_type: 'line',
    },
  ],
  primaryDataset: jest.fn(),
  secondaryDatasets: jest.fn(),
}

export const fakeDataItem = {
  ...fakeDataItemRecordAttrs,
  rawAttributes: jest.fn().mockReturnValue(fakeDataItemRecordAttrs),
  getRecordType: jest.fn().mockReturnValue('items'),
  toJSON: jest.fn().mockReturnValue(fakeDataItemRecordAttrs),
  ...fakeJsonApiAttrs,
}

export const fakeLegendItemAttrs = {
  ...fakeTextItemAttrs,
  type: 'Item::LegendItem',
  primary_measure: {
    measure: 'Business Unit',
    order: 0,
    style: { fill: '#9874AB' },
  },
  legend_search_source: 'search_test_collections',
  primaryDataset: fakeDataset,
  datasets: [
    fakeDataset,
    {
      order: 1,
      measure: '95th Percentile',
      name: '85th',
      fakeDataset,
      selected: true,
    },
    {
      order: 2,
      measure: '75th Percentile',
      name: '75th',
      fakeDataset,
      selected: true,
    },
  ],
}

export const fakeLegendItem = {
  ...fakeLegendItemAttrs,
  rawAttributes: jest.fn().mockReturnValue(fakeLegendItemAttrs),
  getRecordType: jest.fn().mockReturnValue('items'),
  save: jest.fn().mockReturnValue(Promise.resolve({})),
}

export const fakeLegendItemCard = {
  ...fakeItemCard,
  record: fakeLegendItem,
  parent: fakeCollection,
}

export const fakeTextItem = {
  ...fakeTextItemAttrs,
  rawAttributes: jest.fn().mockReturnValue(fakeTextItemAttrs),
  getRecordType: jest.fn().mockReturnValue('items'),
  toJSON: jest.fn().mockReturnValue(fakeTextItemAttrs),
  pushUndo: jest.fn(),
  version: 1,
  ...fakeJsonApiAttrs,
}
export const fakeVideoItemAttrs = {
  id: '2',
  type: 'Item::VideoItem',
  internalType: 'items',
  name: 'Crazy Honey Badger',
  url: 'https://www.youtube.com/watch?v=4r7wHMg5Yjg',
  thumbnail_url: 'https://img.youtube.com/vi/4r7wHMg5Yjg/hqdefault.jpg',
  inherited_tag_list: [],
  can_edit: false,
  parent_collection_card: fakeCollectionCard,
}
export const fakeVideoItem = {
  ...fakeVideoItemAttrs,
  rawAttributes: jest.fn().mockReturnValue(fakeTextItemAttrs),
  getRecordType: jest.fn().mockReturnValue('items'),
}
export const fakeImageItemAttrs = {
  id: '3',
  type: 'Item::FileItem',
  name: 'Earth from Space',
  filestack_file: {
    id: '1',
    url:
      'https://www.nasa.gov/sites/default/files/styles/full_width_feature/public/thumbnails/image/iss052e023801_0.jpg',
  },
  imageUrl: jest.fn(),
  inherited_tag_list: [],
  can_edit: false,
  parent_collection_card: fakeCollectionCard,
}
export const fakeImageItem = {
  ...fakeImageItemAttrs,
  imageUrl: jest.fn().mockReturnValue('http://fake.url/img'),
  rawAttributes: jest.fn().mockReturnValue(fakeTextItemAttrs),
  getRecordType: jest.fn().mockReturnValue('items'),
}
export const fakeLinkItemAttrs = {
  id: '3',
  internalType: 'items',
  type: 'Item::LinkItem',
  name: 'Independent.uk',
  content: 'The best news under the sun',
  icon_url: 'http://icon.jpg',
  thumbnail_url: 'http://thumb.jpg',
  url: 'http://independente.co.uk',
  inherited_tag_list: [],
  can_edit: false,
  parent_collection_card: fakeCollectionCard,
}
export const fakeLinkItem = {
  ...fakeLinkItemAttrs,
  rawAttributes: jest.fn().mockReturnValue(fakeLinkItemAttrs),
  getRecordType: jest.fn().mockReturnValue('items'),
}
export const fakeFileItemAttrs = {
  id: '3',
  type: 'Item::FileItem',
  name: '',
  filestack_file: {
    id: '1',
    url:
      'https://www.nasa.gov/sites/default/files/styles/full_width_feature/public/thumbnails/image/iss052e023801_0.ppt',
    handle: 'aaaaaa',
    mimetype: 'application/vnd.powerpoint',
  },
  inherited_tag_list: [],
  can_edit: false,
  can_view: true,
  parent_collection_card: fakeCollectionCard,
}
export const fakeFileItem = {
  ...fakeFileItemAttrs,
  rawAttributes: jest.fn().mockReturnValue(fakeTextItemAttrs),
  getRecordType: jest.fn().mockReturnValue('items'),
}
export const fakeQuestionItem = {
  id: '3',
  // ...fakeTextItemAttrs,
  type: 'Item::QuestionItem',
  question_type: 'question_open',
  isSingleChoiceQuestion: false,
  rawAttributes: jest.fn().mockReturnValue(fakeTextItemAttrs),
  getRecordType: jest.fn().mockReturnValue('items'),
  API_destroyQuestionChoice: jest.fn().mockReturnValue(Promise.resolve()),
  API_createQuestionChoice: jest.fn().mockReturnValue(Promise.resolve()),
  ...fakeJsonApiAttrs,
}

export const fakeQuestionAnswer = {
  answer_text: 'Great!',
  answer_number: 1,
  question_id: fakeQuestionItem.id,
  selected_choice_ids: [],
}

export const fakeQuestionChoice = {
  text: 'Option A',
  question_item_id: fakeQuestionItem.id,
  order: 0,
  value: '0',
  id: 1,
}

export const fakeQuestionSecondChoice = {
  text: 'Option B',
  question_item_id: fakeQuestionItem.id,
  order: 1,
  value: '1',
  id: 2,
}

export const fakeItemCard = {
  id: '10',
  order: 0,
  height: 1,
  width: 1,
  maxWidth: 1,
  record: fakeTextItem,
  item: fakeTextItem,
  image_contain: false,
  API_create: jest.fn(),
  API_archive: jest.fn(),
  API_linkToMyCollection: jest.fn(),
  ...fakeJsonApiAttrs,
}

export const fakeQuestionItemCard = {
  ...fakeItemCard,
  record: fakeQuestionItem,
  card_question_type: fakeQuestionItem.question_type,
}

const fakeCards = [fakeItemCard, fakeItemCard, fakeItemCard]

export const fakeCollection = {
  id: '1',
  name: 'My Workspace X',
  pageTitle: 'My Workspace X | Shape',
  type: 'Collection',
  breadcrumb: [],
  roles: [],
  tag_list: ['prototype', 'blockchain'],
  inherited_tag_list: [],
  can_edit: false,
  can_view: true,
  can_edit_content: false,
  master_template: false,
  isCollection: true,
  isSharedCollection: false,
  isUserCollection: false,
  isNormalCollection: true,
  num_survey_responses: 0,
  anyone_can_view: false,
  anyone_can_join: false,
  recordsPerPage: 50,
  searchRecordsPerPage: 20,
  updated_at: "2019-11-22T18:57:12.863Z",
  cover: {
    image_url: 'http://fake.url.net',
    text: 'Lorem ipsum blockchain boogie',
    name: 'Proto-typo',
    hardcoded_subtitle: 'Lorem ipsum hardcoded',
    subtitle_hidden: false,
  },
  links: {
    self: 'https://www.shape.space/ideo/collections/1',
  },
  collection_filters: [],
  card_order: 'order',
  collection_cards: fakeCards,
  sortedCards: fakeCards,
  parent_collection_card: fakeCollectionCard,
  // This is a computed property on the collection store
  cardIds: _.map(fakeCards, c => c.id),
  cardIdsWithinRectangle: jest.fn().mockReturnValue([]),
  addCard: jest.fn(),
  API_archive: jest.fn(),
  API_updateCard: jest.fn(),
  API_updateNameAndCover: jest.fn(),
  API_getNextAvailableTest: jest.fn(),
  API_clearCollectionCover: jest.fn(),
  API_fetchCards: jest.fn().mockReturnValue(Promise.resolve({})),
  API_batchUpdateCardsWithUndo: jest.fn().mockReturnValue(Promise.resolve({})),
  API_createCollectionFilter: jest.fn().mockReturnValue(Promise.resolve({})),
  API_destroyCollectionFilter: jest.fn().mockReturnValue(Promise.resolve({})),
  API_selectDatasetsWithIdentifier: jest
    .fn()
    .mockReturnValue(Promise.resolve({})),
  API_unselectDatasetsWithIdentifier: jest
    .fn()
    .mockReturnValue(Promise.resolve({})),
  API_removeComparison: jest.fn().mockReturnValue(Promise.resolve({})),
  API_addComparison: jest.fn().mockReturnValue(Promise.resolve({})),
  API_selectCollectionType: jest.fn().mockReturnValue(Promise.resolve({})),
  checkCurrentOrg: jest.fn(),
  confirmEdit: jest.fn(),
  updateScrollBottom: jest.fn(),
  clearCollectionCards: jest.fn(),
  cardProperties: [],
  internalType: 'collections',
  collection_type: 'method',
  meta: {
    snapshot: {
      can_edit: false,
    },
  },
  ...fakeJsonApiAttrs,
}
// also set parentCollection on fakeCard
// TODO: fix circular reference!
fakeCollectionCard.parentCollection = { ...fakeCollection }

export const fakeGroupAttrs = {
  id: '1',
  name: 'IDEO Products',
  handle: 'ideo-products',
  internalType: 'groups',
  filestack_file_url: 'https://cdn.filestackcontent.com/i4iKADquTQCWMAvyz02R',
  roles: [],
  can_edit: true,
}
export const fakeGroup = {
  ...fakeJsonApiAttrs,
  ...fakeGroupAttrs,
  rawAttributes: jest.fn().mockReturnValue(fakeGroupAttrs),
  getRecordType: jest.fn().mockReturnValue('groups'),
}
export const fakeOrganizationAttrs = {
  id: '1',
  name: 'TestOrg',
  slug: 'test-org',
  primary_group: fakeGroup,
  guest_group: fakeGroup,
  admin_group: fakeGroup,
}
export const fakeOrganization = {
  ...fakeOrganizationAttrs,
  rawAttributes: jest.fn().mockReturnValue(fakeOrganizationAttrs),
  getRecordType: jest.fn().mockReturnValue('organization'),
}
export const fakeUserAttrs = {
  id: '1',
  first_name: 'Oprah',
  last_name: 'Winfrey',
  name: 'Oprah Winfrey',
  email: 'oprah@winfrey.com',
  pic_url_square: 'https://filestackcdn.com/abc123',
  groups: [],
  current_organization: fakeOrganization,
  type: 'users',
  terms_accepted: true,
  isCurrentUser: true,
  feedback_contact_preference: 'feedback_contact_unanswered',
  API_updateCurrentUser: jest.fn(),
  API_updateSurveyRespondent: jest.fn(),
  API_hideHelper: jest.fn().mockReturnValue(Promise.resolve({})),
  notify_through_email: true,
  show_move_modal: false,
}
export const fakeActionCableUser = {
  id: '99',
  name: 'Oprah Winfrey',
  pic_url_square: 'https://filestackcdn.com/abc123',
}
export const fakeUser = {
  ...fakeUserAttrs,
  rawAttributes: jest.fn().mockReturnValue(fakeUserAttrs),
  getRecordType: jest.fn().mockReturnValue('users'),
}
export const fakeRoleAttrs = {
  id: '1',
  name: 'editor',
  users: [fakeUser, fakeUser],
  groups: [],
  resource: { id: '1', type: 'collection', internalType: 'collection' },
  API_delete: jest.fn().mockReturnValue(Promise.resolve({})),
}
export const fakeRole = {
  ...fakeRoleAttrs,
  rawAttributes: jest.fn().mockReturnValue(fakeRoleAttrs),
  getRecordType: jest.fn().mockReturnValue('roles'),
  canEdit: jest.fn(),
}
export const fakePosition = {
  xPos: 0,
  yPos: 0,
  height: 1,
  width: 1,
}
export const fakeComment = {
  id: '1',
  author: fakeUser,
  message:
    'This is my message to the world. Go to the source: https://www.ideo.com',
  draftjs_data: {
    blocks: [
      {
        key: '74h6f',
        data: {},
        text:
          'This is my message to the world. Go to the source: https://www.ideo.com',
        type: 'unstyled',
        depth: 0,
        entityRanges: [],
        inlineStyleRanges: [],
      },
    ],
    entityMap: {},
  },
  replies: [],
  created_at: new Date('2019-05-09T03:18:00'),
  updated_at: new Date('2019-05-09T03:18:00'),
  API_fetchReplies: jest.fn().mockReturnValue(Promise.resolve({})),
  expandAndFetchReplies: jest.fn().mockReturnValue(Promise.resolve({})),
}
export const fakeThread = {
  id: '1',
  persisted: true,
  record: fakeCollection,
  key: 'collection-1',
  unread_count: 2,
  comments: [fakeComment, fakeComment, fakeComment],
  users_thread: {
    subscribed: true,
    get currentSubscribed() {
      return this.subscribed
    },
  },
  visibleCommentsAndRepliesCount: 1,
  API_saveComment: jest.fn().mockReturnValue(Promise.resolve({})),
  API_fetchComments: jest.fn().mockReturnValue(Promise.resolve({})),
  API_subscribe: jest.fn().mockReturnValue(Promise.resolve({})),
  API_unsubscribe: jest.fn().mockReturnValue(Promise.resolve({})),
}

export const fakeActivity = {
  id: '1',
  type: 'activities',
  action: 'archived',
  actor: fakeUser,
  created_at: new Date(),
  subject_users: [],
  subject_groups: [],
  target: fakeCollection,
  target_type: 'Collection',
  target_id: fakeCollection.id,
  setTarget: jest.fn(),
  ...fakeJsonApiAttrs,
}
export const fakeNotification = {
  id: '1',
  type: 'notifications',
  read: false,
  activity: fakeActivity,
  combined_actors: [],
  combined_activities_ids: [],
  user: fakeUser,
  ...fakeJsonApiAttrs,
}
export const fakeSurveyResponse = {
  id: '1',
  type: 'survey_responses',
  status: 'in_progress',
  question_answers: [],
  ...fakeJsonApiAttrs,
}
export const fakeAudience = {
  id: '1',
  name: 'Anybody',
  min_price_per_response: 4.0,
}
export const fakeTestAudience = {
  id: '1',
  audience: fakeAudience,
  sample_size: 12,
  num_completed_responses: 6,
  price_per_response: 4.24,
  incentive_per_response: 1.95,
}
export const fakeTestCollection = {
  id: '1',
  type: 'test_collections',
  name: 'Test Collection',
  test_launched_at: new Date('2019-05-09T03:18:00'),
  test_audiences: [fakeTestAudience],
}

export const fakeCollectionFilter = {
  id: '1',
  filter_type: 'tag',
  text: 'plants',
  selected: true,
  API_toggleSelected: jest.fn().mockReturnValue(Promise.resolve({})),
}
