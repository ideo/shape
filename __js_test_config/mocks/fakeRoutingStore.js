const fakeRoutingStore = {
  previousPageBeforeSearch: null,
  isHomepage: false,
  pathTo: jest.fn().mockReturnValue('/fake/path'),
  routeTo: jest.fn(),
  pathContains: jest.fn(),
  updatePreviousPageBeforeSearch: jest.fn(),
  leaveSearch: jest.fn(),
  slug: jest.fn().mockReturnValue('org-slug'),
  location: {
    pathname: '/xyz',
    search: '',
  }
}

export default fakeRoutingStore
