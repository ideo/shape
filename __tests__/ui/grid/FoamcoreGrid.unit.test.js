import FoamcoreGrid from '~/ui/grid/FoamcoreGrid'

import fakeApiStore from '#/mocks/fakeApiStore'
import fakeUiStore from '#/mocks/fakeUiStore'
import { fakeCollectionCard, fakeCollection } from '#/mocks/data'

let props, wrapper, instance, rerender, cards, cardA, cardB, cardC
let idCounter = 0

function createCard(data) {
  idCounter += 1
  const id = idCounter.toString()
  return { ...fakeCollectionCard, ...data, id }
}

describe('FoamcoreGrid', () => {
  beforeEach(() => {
    cardA = createCard({ col: 5, row: 1 })
    cardB = createCard({ col: 1, row: 0, width: 2, height: 2 })
    cardC = createCard({ col: 0, row: 2, width: 2 })
    const collection = fakeCollection
    collection.collection_cards = [cardA, cardB, cardC]

    props = {
      collection,
      canEditCollection: true,
      gridW: 200,
      gridH: 200,
      gutter: 10,
      sortBy: 'order',
      loadCollectionCards: jest.fn(() => Promise.resolve()),
      updateCollection: jest.fn(),
      cardProperties: [],
      apiStore: fakeApiStore(),
      uiStore: fakeUiStore,
      routingStore: {
        routeTo: jest.fn(),
        push: jest.fn(),
      },
    }
    rerender = () => {
      wrapper = shallow(<FoamcoreGrid.wrappedComponent {...props} />)
      instance = wrapper.instance()
    }
    rerender()
    cards = props.collection.collection_cards
    instance.gridRef = { scrollLeft: 0, scrollTop: 0 }
  })

  describe('calculateFilledSpots', () => {
    it('maps out all the filled spots in the grid as an array', () => {
      expect(instance.filledSpots.length).toEqual(7)
      const firstSpot = { card: cardA, row: 1, col: 5 }
      expect(instance.filledSpots).toContainEqual(firstSpot)

      const secondSpot = { card: cardB, row: 0, col: 1 }
      expect(instance.filledSpots).toContainEqual(secondSpot)

      const lastSpot = { card: cardC, row: 2, col: 1 }
      expect(instance.filledSpots).toContainEqual(lastSpot)
    })
  })

  describe('handleMouseMove', () => {
    const fakeEv = { persist: jest.fn(), pageX: 120, pageY: 50 }

    beforeEach(() => {
      instance.throttledSetHoverSpot = jest.fn().mockReturnValue('')
      instance.handleMouseMove(fakeEv)
    })

    it('should call persist on the event', () => {
      expect(fakeEv.persist).toHaveBeenCalled()
    })

    it('should set the hover spot throttled', () => {
      expect(instance.throttledSetHoverSpot).toHaveBeenCalled()
    })
  })

  describe('handleMouseMove', () => {
    beforeEach(() => {
      instance.placeholderSpot = { row: 1, col: 1 }
      instance.handleMouseOut()
    })

    it('should reset the placeholder spot', () => {
      expect(instance.placeholderSpot.row).toBeUndefined()
      expect(instance.placeholderSpot.col).toBeUndefined()
    })
  })

  describe('onDragStart', () => {
    describe('when dragging multiple cards', () => {
      let fakeDragMap, cardId

      beforeEach(() => {
        fakeDragMap = [
          {
            card: cards[0],
            col: 0,
            row: 0,
          },
          {
            card: cards[1],
            col: 1,
            row: 0,
          },
        ]
        cardId = cards[0].id

        instance.draggingMap = []
        instance.determineDragMap = jest.fn().mockReturnValue(fakeDragMap)
        instance.onDragStart(cardId)
      })

      it('should determine the drag map with card id', () => {
        expect(instance.determineDragMap).toHaveBeenCalledWith(cardId)
      })

      it('should set the dragging map if dragging multiple cards', () => {
        expect(instance.draggingMap).toEqual(fakeDragMap)
      })
    })
  })

  describe('onDragOrResizeStop', () => {
    let cardId

    beforeEach(() => {
      cardId = cards[0].id
      instance.moveCard = jest.fn().mockReturnValue()
      instance.resizeCard = jest.fn().mockReturnValue()
      instance.dragging = true
      // Stub this or else it causes mobx `Maximum call stack size exceeded`
      instance.calculateCardsToRender = jest.fn()
      props.uiStore.multiMoveCardIds = [cards[0].id]
      instance.onDragOrResizeStop(cardId, 'move')
    })

    it('should stop all dragging', () => {
      expect(instance.dragGridSpot.size).toEqual(0)
      expect(instance.dragging).toEqual(false)
      expect(props.uiStore.multiMoveCardIds.length).toBe(0)
      expect(instance.calculateCardsToRender).toHaveBeenCalled()
    })

    describe('when moving', () => {
      beforeEach(() => {
        instance.onDragOrResizeStop(cardId, 'move')
      })

      it('should try and move the card with the found card', () => {
        expect(instance.moveCard).toHaveBeenCalledWith(cards[0])
        expect(instance.calculateCardsToRender).toHaveBeenCalled()
      })
    })

    describe('when resizing', () => {
      beforeEach(() => {
        instance.resizing = true
        instance.onDragOrResizeStop(cardId, 'resize')
      })

      it('should try and resize the card with the found card', () => {
        expect(instance.resizeCard).toHaveBeenCalledWith(cards[0])
        expect(instance.calculateCardsToRender).toHaveBeenCalled()
      })

      it('should stop all resizing and moving', () => {
        expect(instance.resizing).toEqual(false)
        expect(instance.calculateCardsToRender).toHaveBeenCalled()
      })
    })
  })

  describe('onResize', () => {
    let cardId

    beforeEach(() => {
      instance.resizing = false
      cardId = cards[0].id
      instance.throttledSetResizeSpot = jest.fn().mockReturnValue()
      instance.onResize(cardId, { width: 2, height: 1 })
    })

    it('should set resizing to true', () => {
      expect(instance.resizing).toBe(true)
    })

    it('should call to set the resize spot with dimensions and position', () => {
      expect(instance.throttledSetResizeSpot).toHaveBeenCalledWith({
        row: cards[0].row,
        col: cards[0].col,
        width: 2,
        height: 1,
      })
    })
  })

  describe('calcEdgeCol/Row', () => {
    beforeEach(() => {
      cardA = createCard({ col: 1, row: 1 })
      cardB = createCard({ col: 8, row: 1 })
      cardC = createCard({ col: 9, row: 9 })
      props.collection.collection_cards = [cardA, cardB, cardC]
    })

    describe('with a card that has no cards around it', () => {
      it('should set the max column as the max card width', () => {
        const edgeCol = instance.calcEdgeCol(cards[0], cards[0].id)
        expect(edgeCol).toEqual(4)
      })

      it('sets the max row as the max card height', () => {
        const edgeRow = instance.calcEdgeRow(cards[0], cards[0].id)
        expect(edgeRow).toEqual(2)
      })
    })

    describe('with a card that has horizontal contraints, 2 spaces apart', () => {
      beforeEach(() => {
        const otherCard = props.collection.collection_cards[1]
        otherCard.row = 1
        otherCard.col = 3
        rerender()
      })

      it('has a maxResizeCol of 2', () => {
        const edgeCol = instance.calcEdgeCol(cardA, cardA.id)
        expect(edgeCol).toEqual(2)
      })
    })

    describe('with a card that has vertical contraints, 1 space apart', () => {
      beforeEach(() => {
        const otherCard = props.collection.collection_cards[1]
        otherCard.row = 2
        otherCard.col = 1
        rerender()
      })

      it('has maxResizeRow of 1', () => {
        const edgeRow = instance.calcEdgeRow(cardA, cardA.id)
        expect(edgeRow).toEqual(1)
      })
    })
  })

  describe('loadAfterScroll', () => {
    beforeEach(() => {
      instance.loadedRows = { min: 0, max: 9 }
      instance.loadedCols = { min: 0, max: 9 }
      instance.loadCards = jest.fn()
    })

    describe('scrolling in loaded bounds', () => {
      beforeEach(() => {
        // `Object.defineProperty` is the only way I could find to stub getter methods
        Object.defineProperty(instance, 'visibleCols', {
          get: jest.fn().mockReturnValue({ min: 0, max: 4, num: 5 }),
        })
        Object.defineProperty(instance, 'visibleRows', {
          get: jest.fn().mockReturnValue({ min: 1, max: 4, num: 4 }),
        })
      })

      it('does not call loadCards if all in view', () => {
        instance.loadAfterScroll()
        expect(instance.loadCards).not.toHaveBeenCalled()
      })
    })

    describe('scrolling out of bounds vertically', () => {
      beforeEach(() => {
        Object.defineProperty(instance, 'visibleRows', {
          get: jest.fn().mockReturnValue({ min: 4, max: 8, num: 4 }),
        })
        Object.defineProperty(instance, 'visibleCols', {
          get: jest.fn().mockReturnValue({ min: 0, max: 4, num: 5 }),
        })
      })

      it('calls loadMoreRows', () => {
        const expectedRowsCols = {
          cols: [0, 10],
          rows: [10, 14],
        }
        instance.loadAfterScroll()
        expect(instance.loadCards).toHaveBeenCalledWith(expectedRowsCols)
      })
    })

    describe('scrolling out of bounds horizontally', () => {
      beforeEach(() => {
        Object.defineProperty(instance, 'visibleCols', {
          get: jest.fn().mockReturnValue({ min: 2, max: 6, num: 5 }),
        })
        Object.defineProperty(instance, 'visibleRows', {
          get: jest.fn().mockReturnValue({ min: 0, max: 3, num: 4 }),
        })
      })

      it('calls loadMoreRows', () => {
        const expectedRowsCols = {
          cols: [10, 15],
          rows: [0, 8],
        }
        instance.loadAfterScroll()
        expect(instance.loadCards).toHaveBeenCalledWith(expectedRowsCols)
      })
    })
  })

  describe('loadCards', () => {
    beforeEach(() => {
      props.loadCollectionCards = jest.fn()
      rerender()
    })

    it('calls props.loadCollectionCards', () => {
      const rowsCols = {
        cols: [10, 15],
        rows: [0, 8],
      }
      instance.loadCards(rowsCols)
      expect(props.loadCollectionCards).toHaveBeenCalledWith(rowsCols)
    })
  })

  describe('determineDragMap', () => {})
  describe('resizeCard', () => {})
  describe('moveCard', () => {})
  describe('setResizeSpot', () => {})
})
