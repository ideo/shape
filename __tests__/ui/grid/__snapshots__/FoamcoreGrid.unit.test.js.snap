// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FoamcoreGrid renders snapshot 1`] = `
<styled.div
  data-empty-space-click={true}
  height={NaN}
  innerRef={[Function]}
  onScroll={[Function]}
  width={988.2790697674419}
>
  <FoamcoreZoomControls
    onZoomIn={[Function]}
    onZoomOut={[Function]}
  />
  <styled.div
    blocked={false}
    data-blank-type="unrendered"
    data-empty-space-click={true}
    draggedOn={true}
    h={250}
    height={250}
    key="blank-unrendered-1:5"
    onClick={[Function]}
    type="unrendered"
    w={312}
    width={312}
    x={308.83720930232556}
    xPos={308.83720930232556}
    y={50.23255813953488}
    yPos={50.23255813953488}
    zoomLevel={5.375}
  >
    <InlineLoader
      background="#f5f4f3"
      fixed={false}
    />
  </styled.div>
  <styled.div
    blocked={false}
    data-blank-type="unrendered"
    data-empty-space-click={true}
    draggedOn={true}
    h={520}
    height={520}
    key="blank-unrendered-0:1"
    onClick={[Function]}
    type="unrendered"
    w={644}
    width={644}
    x={61.76744186046512}
    xPos={61.76744186046512}
    y={0}
    yPos={0}
    zoomLevel={5.375}
  >
    <InlineLoader
      background="#f5f4f3"
      fixed={false}
    />
  </styled.div>
  <styled.div
    blocked={false}
    data-blank-type="unrendered"
    data-empty-space-click={true}
    draggedOn={true}
    h={250}
    height={250}
    key="blank-unrendered-2:0"
    onClick={[Function]}
    type="unrendered"
    w={644}
    width={644}
    x={0}
    xPos={0}
    y={100.46511627906976}
    yPos={100.46511627906976}
    zoomLevel={5.375}
  >
    <InlineLoader
      background="#f5f4f3"
      fixed={false}
    />
  </styled.div>
</styled.div>
`;
